src/util/worker/mod.rs

```diff
@@ -22,94 +22,248 @@ quick_error! {
     }
 }
 
+impl<T> From<mpsc::SendError<T>> for Error {
+    fn from(_: mpsc::SendError<T>) -> Error {
+        // do we need to return the failed data
+        Error::Stopped
+    }
+}
+
 pub type Result<T> = result::Result<T, Error>;
 
 pub trait Runnable<T: Display> {
     fn run(&mut self, t: T);
 }
 
+pub trait BatchRunnable<T: Display> {
+    fn run_batch(&mut self, ts: Vec<T>);
+}
+
+impl<T: Display, R: Runnable<T>> BatchRunnable<T> for R {
+    fn run_batch(&mut self, ts: Vec<T>) {
+        for t in ts {
+            let task_str = format!("{}", t);
+            let timer = SlowTimer::new();
+            self.run(t);
+            slow_log!(timer,
+                      "task {} takes {:?} to finish.",
+                      task_str,
+                      timer.elapsed());
+        }
+    }
+}
+
+/// Scheduler provide interface to schedule task to underlying workers.
+pub struct Scheduler<T> {
+    counter: Arc<AtomicUsize>,
+    sender: Sender<Option<T>>,
+}
+
+impl<T: Display> Scheduler<T> {
+    fn new(counter: AtomicUsize, sender: Sender<Option<T>>) -> Scheduler<T> {
+        Scheduler {
+            counter: Arc::new(counter),
+            sender: sender,
+        }
+    }
+
+    /// Schedule a task to run.
+    ///
+    /// If the worker is stopped, an error will return.
+    pub fn schedule(&self, task: T) -> Result<()> {
+        debug!("scheduling task {}", task);
+        try!(self.sender.send(Some(task)));
+        self.counter.fetch_add(1, Ordering::SeqCst);
+        Ok(())
+    }
+
+    /// Check if underlying worker can't handle task immediately.
+    pub fn is_busy(&self) -> bool {
+        self.counter.load(Ordering::SeqCst) > 0
+    }
+}
+
+impl<T: Display> Clone for Scheduler<T> {
+    fn clone(&self) -> Scheduler<T> {
+        Scheduler {
+            counter: self.counter.clone(),
+            sender: self.sender.clone(),
+        }
+    }
+}
+
 /// A worker that can schedule time consuming tasks.
 pub struct Worker<T: Display> {
     name: String,
-    counter: Arc<AtomicUsize>,
-    sender: Option<Sender<T>>,
-    receiver: Option<Receiver<T>>,
+    batch_size: usize,
+    scheduler: Option<Scheduler<T>>,
+    receiver: Option<Receiver<Option<T>>>,
```

Comment:
why using Option here? seem you can use thread handle to check worker started or not too.